File: Dart/cnpj.dart
import 'dart:math';

/// Classe para validação e cálculo de dígitos verificadores de CNPJs alfanuméricos
class CNPJ {
  /// Tamanho do CNPJ sem os dígitos verificadores
  static const int tamanhoCnpjSemDv = 12;
  
  /// Tamanho do CNPJ completo
  static const int tamanhoCnpjCompleto = 14;
  
  /// Padrão Regex para validar CNPJs alfanuméricos completos
  static final RegExp regexCnpj = RegExp(r'^([A-Z\d]){12}(\d){2}$');
  
  /// Padrão Regex para validar bases de CNPJs alfanuméricos (sem DVs)
  static final RegExp regexCnpjSemDv = RegExp(r'^([A-Z\d]){12}$');
  
  /// Padrão Regex para identificar caracteres de máscara
  static final RegExp regexCaracteresMascara = RegExp(r'[./-]');
  
  /// Padrão Regex para validar caracteres permitidos (incluindo máscara)
  static final RegExp regexCaracteresPermitidos = RegExp(r'^[A-Z\d./-]*$');
  
  /// Pesos para o cálculo do primeiro dígito verificador
  static const List<int> pesosDv1 = [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];
  
  /// Pesos para o cálculo do segundo dígito verificador
  static const List<int> pesosDv2 = [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2];

  /// Verifica se um CNPJ alfanumérico é válido
  /// 
  /// Retorna `true` se o CNPJ é válido, ou `false` caso contrário
  static bool isValid(String cnpj) {
    // Verifica se o CNPJ contém apenas caracteres permitidos
    if (!regexCaracteresPermitidos.hasMatch(cnpj)) {
      return false;
    }

    // Remove máscara e converte para maiúsculas
    String cnpjSemMascara = removeMascaraCnpj(cnpj.toUpperCase());
    
    // Verifica o tamanho do CNPJ sem máscara
    if (cnpjSemMascara.length != tamanhoCnpjCompleto) {
      return false;
    }
    
    // Verifica se o CNPJ não contém apenas zeros
    if (apenasZeros(cnpjSemMascara)) {
      return false;
    }
    
    // Verifica se o CNPJ corresponde ao padrão correto
    if (!regexCnpj.hasMatch(cnpjSemMascara)) {
      return false;
    }
    
    // Extrai os dígitos verificadores informados
    String dvInformado = cnpjSemMascara.substring(tamanhoCnpjSemDv);
    
    // Calcula os dígitos verificadores
    try {
      String dvCalculado = calculaDV(cnpjSemMascara.substring(0, tamanhoCnpjSemDv));
      
      // Compara os dígitos verificadores
      return dvInformado == dvCalculado;
    } catch (e) {
      return false;
    }
  }

  /// Calcula os dígitos verificadores de um CNPJ alfanumérico
  /// 
  /// Retorna uma string contendo os dois dígitos verificadores
  /// ou lança uma exceção em caso de erro
  static String calculaDV(String cnpj) {
    // Verifica se o CNPJ contém apenas caracteres permitidos
    if (!regexCaracteresPermitidos.hasMatch(cnpj)) {
      throw ArgumentError('CNPJ contém caracteres não permitidos');
    }
    
    // Remove máscara e converte para maiúsculas
    String cnpjSemMascara = removeMascaraCnpj(cnpj.toUpperCase());
    
    // Usa apenas os primeiros 12 caracteres
    if (cnpjSemMascara.length >= tamanhoCnpjSemDv) {
      cnpjSemMascara = cnpjSemMascara.substring(0, tamanhoCnpjSemDv);
    }
    
    // Verifica o tamanho da base do CNPJ
    if (cnpjSemMascara.length != tamanhoCnpjSemDv) {
      throw ArgumentError('CNPJ deve ter $tamanhoCnpjSemDv caracteres para calcular o DV');
    }
    
    // Verifica se o CNPJ corresponde ao padrão correto
    if (!regexCnpjSemDv.hasMatch(cnpjSemMascara)) {
      throw ArgumentError('CNPJ não está no formato correto');
    }
    
    // Verifica se o CNPJ não contém apenas zeros
    if (apenasZeros(cnpjSemMascara)) {
      throw ArgumentError('CNPJ não pode conter apenas zeros');
    }
    
    // Calcula o primeiro dígito verificador
    int dv1 = calcularDigito(cnpjSemMascara, pesosDv1);
    
    // Adiciona o primeiro dígito verificador à base para calcular o segundo
    String cnpjComDv1 = cnpjSemMascara + dv1.toString();
    
    // Calcula o segundo dígito verificador
    int dv2 = calcularDigito(cnpjComDv1, pesosDv2);
    
    // Retorna os dígitos verificadores como string
    return '$dv1$dv2';
  }

  /// Remove a máscara de formatação do CNPJ
  /// 
  /// Retorna o CNPJ sem caracteres de formatação
  static String removeMascaraCnpj(String cnpj) {
    return cnpj.replaceAll(regexCaracteresMascara, '');
  }

  /// Converte um caractere alfanumérico para seu valor numérico
  /// 
  /// Retorna o valor numérico equivalente ao caractere
  static int converterCaractere(String caractere) {
    // Pega o código ASCII do caractere
    int ascii = caractere.codeUnitAt(0);
    
    // Se for dígito, retorna seu valor numérico
    if (ascii >= 48 && ascii <= 57) {
      return ascii - 48;
    }
    // Se for letra maiúscula, retorna seu valor convertido
    else if (ascii >= 65 && ascii <= 90) {
      return ascii - 48 - 7; // Convertendo para A=10, B=11, etc.
    }
    // Caractere inválido
    throw ArgumentError('Caractere inválido: $caractere');
  }

  /// Verifica se um CNPJ contém apenas zeros
  /// 
  /// Retorna `true` se o CNPJ contiver apenas zeros, `false` caso contrário
  static bool apenasZeros(String cnpj) {
    return RegExp(r'^0+$').hasMatch(cnpj);
  }

  /// Calcula um dígito verificador com base nos pesos fornecidos
  /// 
  /// Retorna o dígito verificador calculado
  static int calcularDigito(String cnpj, List<int> pesos) {
    if (cnpj.length > pesos.length) {
      throw ArgumentError('Tamanho do CNPJ incompatível com os pesos');
    }
    
    int soma = 0;
    
    // Calcula a soma dos produtos
    for (int i = 0; i < cnpj.length; i++) {
      int valor = converterCaractere(cnpj[i]);
      soma += valor * pesos[i];
    }
    
    // Calcula o dígito verificador
    int resto = soma % 11;
    if (resto < 2) {
      return 0;
    } else {
      return 11 - resto;
    }
  }
}

File: C/cnpj.c
#include "cnpj.h"
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <stdlib.h>

#define TAMANHO_CNPJ_SEM_DV 12
#define TAMANHO_CNPJ_COMPLETO 14
#define TAMANHO_BUFFER 20

static const int pesos_dv1[] = {5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};
static const int pesos_dv2[] = {6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2};

// Converte um caractere alfanumérico para seu valor numérico
static int converter_caractere(char c) {
    // Se for dígito, retorna seu valor numérico
    if (isdigit(c)) {
        return c - '0';
    }
    // Se for letra maiúscula, retorna seu valor convertido (A=17, B=18, etc.)
    else if (isupper(c)) {
        return c - '0' - 7; // 'A' ASCII 65 - '0' ASCII 48 - 7 = 10
    }
    // Caractere inválido
    return -1;
}

// Verifica se um caractere é permitido (letra maiúscula ou dígito)
static bool caractere_permitido(char c) {
    return (isdigit(c) || (isupper(c) && c >= 'A' && c <= 'Z'));
}

// Verifica se uma string contém apenas caracteres permitidos
static bool apenas_caracteres_permitidos(const char* str) {
    while (*str) {
        if (*str != '.' && *str != '/' && *str != '-' && !caractere_permitido(*str)) {
            return false;
        }
        str++;
    }
    return true;
}

// Verifica se um CNPJ contém apenas zeros
static bool apenas_zeros(const char* cnpj) {
    for (int i = 0; i < TAMANHO_CNPJ_COMPLETO && cnpj[i]; i++) {
        if (cnpj[i] != '0') {
            return false;
        }
    }
    return true;
}

// Calcula um dígito verificador com base nos pesos fornecidos
static int calcular_digito(const char* cnpj, const int* pesos, int tamanho) {
    int soma = 0;
    
    for (int i = 0; i < tamanho; i++) {
        int valor = converter_caractere(cnpj[i]);
        if (valor < 0) return -1; // Caractere inválido
        soma += valor * pesos[i];
    }
    
    int resto = soma % 11;
    if (resto < 2) {
        return 0;
    } else {
        return 11 - resto;
    }
}

void remove_mascara_cnpj(const char* cnpj, char* cnpj_sem_mascara, int tamanho) {
    int j = 0;
    for (int i = 0; cnpj[i] && j < tamanho - 1; i++) {
        if (cnpj[i] != '.' && cnpj[i] != '/' && cnpj[i] != '-') {
            cnpj_sem_mascara[j++] = toupper(cnpj[i]);
        }
    }
    cnpj_sem_mascara[j] = '\0';
}

bool is_valid(const char* cnpj) {
    if (!cnpj || !apenas_caracteres_permitidos(cnpj)) {
        return false;
    }
    
    char cnpj_sem_mascara[TAMANHO_BUFFER];
    remove_mascara_cnpj(cnpj, cnpj_sem_mascara, TAMANHO_BUFFER);
    
    // Verifica o tamanho do CNPJ sem máscara
    if (strlen(cnpj_sem_mascara) != TAMANHO_CNPJ_COMPLETO) {
        return false;
    }
    
    // Verifica se o CNPJ não contém apenas zeros
    if (apenas_zeros(cnpj_sem_mascara)) {
        return false;
    }
    
    // Extrai os dígitos verificadores informados
    char dv_informado[3] = {cnpj_sem_mascara[TAMANHO_CNPJ_SEM_DV], 
                           cnpj_sem_mascara[TAMANHO_CNPJ_SEM_DV + 1], '\0'};
    
    // Calcula os dígitos verificadores
    char dv_calculado[3];
    if (calcula_dv(cnpj_sem_mascara, dv_calculado) != 0) {
        return false;
    }
    
    // Compara os dígitos verificadores
    return strcmp(dv_informado, dv_calculado) == 0;
}

int calcula_dv(const char* cnpj, char* dv) {
    if (!cnpj || !dv || !apenas_caracteres_permitidos(cnpj)) {
        return -1;
    }
    
    char cnpj_sem_mascara[TAMANHO_BUFFER];
    remove_mascara_cnpj(cnpj, cnpj_sem_mascara, TAMANHO_BUFFER);
    
    // Verifica o tamanho da base do CNPJ
    size_t tamanho = strlen(cnpj_sem_mascara);
    if (tamanho != TAMANHO_CNPJ_SEM_DV && tamanho != TAMANHO_CNPJ_COMPLETO) {
        return -1;
    }
    
    // Usa apenas os primeiros 12 caracteres
    char cnpj_base[TAMANHO_CNPJ_SEM_DV + 1];
    strncpy(cnpj_base, cnpj_sem_mascara, TAMANHO_CNPJ_SEM_DV);
    cnpj_base[TAMANHO_CNPJ_SEM_DV] = '\0';
    
    // Verifica se o CNPJ não contém apenas zeros
    if (apenas_zeros(cnpj_base)) {
        return -1;
    }
    
    // Calcula o primeiro dígito verificador
    int dv1 = calcular_digito(cnpj_base, pesos_dv1, TAMANHO_CNPJ_SEM_DV);
    if (dv1 < 0) {
        return -1;
    }
    
    // Adiciona o primeiro dígito verificador à base para calcular o segundo
    char cnpj_com_dv1[TAMANHO_CNPJ_SEM_DV + 2];
    sprintf(cnpj_com_dv1, "%s%d", cnpj_base, dv1);
    
    // Calcula o segundo dígito verificador
    int dv2 = calcular_digito(cnpj_com_dv1, pesos_dv2, TAMANHO_CNPJ_SEM_DV + 1);
    if (dv2 < 0) {
        return -1;
    }
    
    // Retorna os dígitos verificadores como string
    sprintf(dv, "%d%d", dv1, dv2);
    return 0;
}
